packages/base-map/src/index.js

```diff
@@ -0,0 +1,259 @@
+
+    const center = mapConfig &&
+      mapConfig.initLat &&
+      mapConfig.initLon && [mapConfig.initLat, mapConfig.initLon];
+
+    return (
+      <Map
+        ref="map"
+        className="map"
+        center={center}
+        zoom={mapConfig.initZoom || 13}
+        maxZoom={mapConfig.maxZoom}
+        // onClick={this.onLeftClick}
+        // Note: Map-click is handled via single-click plugin, set up in componentDidMount()
+        onOverlayAdd={onOverlayAdded}
+        onOverlayRemove={onOverlayRemoved}
+        onViewportChanged={onViewportChanged}
+      >
+        {/* Create the layers control, including base map layers and any
+         * user-controlled overlays. */}
+        <LayersControl position="topright">
+          {/* base layers */}
+          {baseLayers &&
+            baseLayers.map((layer, i) => {
+              return (
+                <LayersControl.BaseLayer
+                  name={layer.name}
+                  checked={i === 0}
+                  key={i}
+                >
+                  {
+                    // Fix tile size/zoom offset: https://stackoverflow.com/a/37043490/915811
+                    // Also, split the declaration between the two
+                    // outcomes to avoid error 'attempted to load an infinite number of tiles'
+                    // if the tiles cannot be loaded for any reason.
+                  }
+                  {L.Browser.retina && layer.hasRetinaSupport ? (
+                    <TileLayer
+                      url={layer.url}
+                      attribution={layer.attribution}
+                      maxZoom={layer.maxZoom}
+                      tileSize={512}
+                      zoomOffset={-1}
+                      detectRetina
+                    />
+                  ) : (
+                    <TileLayer
+                      url={layer.url}
+                      attribution={layer.attribution}
+                      maxZoom={layer.maxZoom}
+                      detectRetina
+                    />
+                  )}
+                </LayersControl.BaseLayer>
+              );
+            })}
+
+          {/* user-controlled overlay layers (e.g., vehicle locations, stops) */}
+          {userControlledOverlays.map((child, i) => (
+            <LayersControl.Overlay
+              key={i}
+              name={child.props.name}
+              checked={child.props.visible}
+            >
+              {child}
+            </LayersControl.Overlay>
+          ))}
+        </LayersControl>
+
+        {/* Add the fixed, i.e. non-user-controllable, overlays (e.g., itinerary overlay) */}
+        {fixedOverlays}
+
+        {/* Add the location selection popup and content, if both are set. */}
+        {popupLocation && popupContent && (
+          <Popup position={popupLocation} onClose={onPopupClosed}>
+            {popupContent}
+          </Popup>
+        )}
+      </Map>
+    );
+  }
+}
+
+BaseMap.propTypes = {
+  children: PropTypes.oneOfType([
```

Comment:
Needs documentation of how to define a layer. Also, are these propTypes right?

Reply:
The children prop types do not result in validation errors at least. That takes care of zero, one, or multiple children. The overlays and map contents seem to resolve to type Object as opposed to a more precise type.